sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java

```diff
@@ -0,0 +1,187 @@
+import io.crate.data.BatchIterator;
+import io.crate.data.CloseAssertingBatchIterator;
+import io.crate.data.Row;
+import io.crate.data.Row1;
+import io.crate.executor.transport.ShardUpsertRequest;
+import io.crate.operation.Input;
+import io.crate.operation.collect.CollectExpression;
+import io.crate.operation.collect.RowShardResolver;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.bulk.BulkShardProcessor;
+
+import java.util.BitSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+public class IndexWriterCountBatchIterator implements BatchIterator {
+
+    private final Input<BytesRef> sourceInput;
+    private final RowShardResolver rowShardResolver;
+    private final Supplier<String> indexNameResolver;
+    private final Iterable<? extends CollectExpression<Row, ?>> collectExpressions;
+    private final BulkShardProcessor<ShardUpsertRequest> bulkShardProcessor;
+    private final BatchIterator source;
+    private CompletableFuture<BitSet> loading;
+    private Row currentRow = OFF_ROW;
+    private BitSet result;
+
+    private IndexWriterCountBatchIterator(BatchIterator source,
+                                          Supplier<String> indexNameResolver,
+                                          Input<BytesRef> sourceInput,
+                                          Iterable<? extends CollectExpression<Row, ?>> collectExpressions,
+                                          RowShardResolver rowShardResolver,
+                                          BulkShardProcessor bulkShardProcessor) {
+        this.indexNameResolver = indexNameResolver;
+        this.collectExpressions = collectExpressions;
+        this.sourceInput = sourceInput;
+        this.rowShardResolver = rowShardResolver;
+        this.bulkShardProcessor = bulkShardProcessor;
+        this.source = source;
+    }
+
+    public static BatchIterator newInstance(BatchIterator source, Supplier<String> indexNameResolver,
+                                            Input<BytesRef> sourceInput,
+                                            Iterable<? extends CollectExpression<Row, ?>> collectExpressions,
+                                            RowShardResolver rowShardResolver,
+                                            BulkShardProcessor bulkShardProcessor) {
+        IndexWriterCountBatchIterator delegate = new IndexWriterCountBatchIterator(source, indexNameResolver,
+            sourceInput, collectExpressions, rowShardResolver, bulkShardProcessor);
+        return new CloseAssertingBatchIterator(delegate);
+    }
+
+    @Override
+    public void moveToStart() {
+        raiseIfLoading();
+        currentRow = OFF_ROW;
+        result = null;
+    }
+
+    @Override
+    public boolean moveNext() {
+        if (loading == null) {
+            return false;
+        }
+        raiseIfLoading();
+
+        if (currentRow.equals(OFF_ROW) && result == null) {
```

Comment:
How can `result == null` ever be true here? There is a check couple of lines above which causes moveNext to return early